extensions/mssql/src/contracts.ts

```diff
@@ -36,19 +36,48 @@ export interface RequestSecurityTokenParams {
 	authority: string;
 	provider: string;
 	resource: string;
+	accountId: string;
 	scope: string;
 }
 
 export interface RequestSecurityTokenResponse {
 	accountKey: string;
 	token: string;
+	expiresOn: Number;
 }
 
 export namespace SecurityTokenRequest {
 	export const type = new RequestType<RequestSecurityTokenParams, RequestSecurityTokenResponse, void, void>('account/securityTokenRequest');
 }
 // ------------------------------- </ Security Token Request > ------------------------------------------
 
+// ------------------------------- < Refresh Token Request > ---------------------------------
+
+export interface RefreshTokenNotificationParams {
+	authority: string;
+	provider: string;
+	resource: string;
+	accountId: string;
+	uri: string;
+}
+
+export namespace RefreshTokenNotification {
+	export const type = new NotificationType<RefreshTokenNotificationParams, void>('account/refreshTokenNotification');
```

Comment:
Why are there two notifications? This seems like it should be a request - the caller requests that a token is refreshed and then the response contains the updated token data.

Reply:
And how is this different than the securityTokenRefresh from above?

Reply:
The request wasn't working, @kburtram tried to help me but we were not able to get it working where we were sending a request from STS -> ADS and then waiting for the response.

Reply:
the request was getting sent from STS -> ADS, but the response was not getting sent back.  I could see the response in the message writer from ADS getting sent, but somehow STS was still not receiving the message.

Reply:
Are you sure it was never receiving the message at all? As in did you debug into the message dispatcher loop that processes incoming messages and verify that you never saw it come in over the wire? Karl and I discussed about how the dispatcher loop is currently synchronous which is likely the source of the problem but we should verify that's actually the case here. 

And then one of the suggestions I had was to have STS have a separate loop for sending messages that runs independently of the incoming message loop. Did you look into other options such as that?

Reply:
Yes we debugged in the message dispatcher loop.  I went through a debugging session with Karl, but no matter what we tried, we couldn't get any kind of response from ADS.  Karl suggested I do it this way, which is why I've implemented it the way I did

Reply:
I'm not sure about having STS have a separate loop for sending messages, how can I do that?

Reply:
Bit too complex to just tell you outright - but the basic premise is that currently we have a single dispatcher that handles the incoming and outgoing messages (MessageDispatcher). So we'd want to look into modifying that to be able to handle sending and receiving messages on its own loop.

Honestly though that kind of work is going to be a pretty big refactor - and at that point I would say we should rather be focusing on looking into making the dispatcher loop non-blocking. Which is likely going to be a fairly big task in itself given what Karl and Amir found when investigating this a while back.

So this approach is fine for now, although I'd like to see a bit more investigation to make sure that it is indeed the same issue. Starting with my question above about verifying whether we actually get the messages - you should look at https://github.com/microsoft/sqltoolsservice/blob/main/src/Microsoft.SqlTools.Hosting/Hosting/Protocol/MessageDispatcher.cs#L259. Knowing if this is indeed causing the issue will help us determine how big of an issue this is and prioritize it correctly.